src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -318,22 +486,80 @@ internal void NotifyWorkspaceHosts(Action<IVisualStudioWorkspaceHost> action)
             }
         }
 
+        /// <summary>
+        /// Attempts to get single project by given output binary filePath.
+        /// </summary>
+        /// <remarks>This method may be called on a background thread.</remarks>
         internal bool TryGetProjectByBinPath(string filePath, out AbstractProject project)
         {
-            project = null;
+            lock (_projectsByBinPath)
```

Comment:
Shouldn't this be lock (_gate)? Even if not, I'd much prefer a single gate be used for everything instead of separate gates for separate items.